server/packetio.go

```diff
@@ -147,43 +154,58 @@ func (p *packetIO) writePacket(data []byte) error {
 	length := len(data) - 4
 	writePacketBytes.Observe(float64(len(data)))
 
-	for length >= mysql.MaxPayloadLen {
-		data[0] = 0xff
-		data[1] = 0xff
-		data[2] = 0xff
-
-		data[3] = p.sequence
-
-		if n, err := p.bufWriter.Write(data[:4+mysql.MaxPayloadLen]); err != nil {
-			return errors.Trace(mysql.ErrBadConn)
-		} else if n != (4 + mysql.MaxPayloadLen) {
-			return errors.Trace(mysql.ErrBadConn)
-		} else {
-			p.sequence++
-			length -= mysql.MaxPayloadLen
-			data = data[mysql.MaxPayloadLen:]
-		}
-	}
+	w := p.Writer()
+	// now we can ensure the length can't exceed maxpayload
+	// for length >= mysql.MaxPayloadLen {
+	// 	data[0] = 0xff
+	// 	data[1] = 0xff
+	// 	data[2] = 0xff
+
+	// 	data[3] = p.sequence
+
+	// 	buf, err := w.Malloc(len(data[:4+mysql.MaxPayloadLen]))
+	// 	if err != nil {
+	// 		return errors.Trace(mysql.ErrBadConn)
+	// 	}
+	// 	copy(buf, data[:4+mysql.MaxPayloadLen])
+	// 	// if n, err := p.bufWriter.Write(data[:4+mysql.MaxPayloadLen]); err != nil {
+	// 	// 	return errors.Trace(mysql.ErrBadConn)
+	// 	// } else if n != (4 + mysql.MaxPayloadLen) {
+	// 	// 	return errors.Trace(mysql.ErrBadConn)
+	// 	// } else {
+	// 	p.sequence++
+	// 	length -= mysql.MaxPayloadLen
+	// 	data = data[mysql.MaxPayloadLen:]
+	// 	//}
+	// }
 
 	data[0] = byte(length)
 	data[1] = byte(length >> 8)
 	data[2] = byte(length >> 16)
 	data[3] = p.sequence
 
-	if n, err := p.bufWriter.Write(data); err != nil {
-		terror.Log(errors.Trace(err))
-		return errors.Trace(mysql.ErrBadConn)
-	} else if n != len(data) {
+	buf, err := w.Malloc(len(data))
```

Comment:
If `data` is not reused before `w.Flush`, may use `w.WriteBinary(data)` instead of `w.Malloc && copy` to reduce copy